torch/nn/modules/module.py

```diff
@@ -551,6 +551,81 @@ def __repr__(self):
         tmpstr = tmpstr + ')'
         return tmpstr
 
+    def summary(self, input_size):
+        def register_hook(module):
+            def hook(module, input, output):
+                class_name = str(module.__class__).split('.')[-1].split("'")[0]
+                module_idx = len(summary)
+                m_key = '%s-%i' % (class_name, module_idx + 1)
+                summary[m_key] = OrderedDict()
+                summary[m_key]['input_shape'] = list(input[0].size())
+                summary[m_key]['input_shape'][0] = None
+                summary[m_key]['output_shape'] = list(output.size())
+                summary[m_key]['output_shape'][0] = None
+
+                params = 0
+                if hasattr(module, 'weight'):
+                    params += torch.prod(torch.LongTensor(list(module.weight.size())))
+                    if module.weight.requires_grad:
+                        summary[m_key]['trainable'] = True
+                    else:
+                        summary[m_key]['trainable'] = False
+                if hasattr(module, 'bias'):
+                    params += torch.prod(torch.LongTensor(list(module.bias.size())))
+                summary[m_key]['nb_params'] = params
+
+            if not isinstance(module, torch.nn.Sequential) and \
+               not isinstance(module, torch.nn.ModuleList) and \
+               not (module == self):
+                hooks.append(module.register_forward_hook(hook))
+
+        # check if there are multiple inputs to the network
+        if isinstance(input_size[0], (list, tuple)):
+            x = [Variable(th.rand(1, *in_size)) for in_size in input_size]
+        else:
+            x = Variable(torch.randn(1, *input_size))
+
+        # create properties
+        summary = OrderedDict()
+        hooks = []
+        # register hook
+        self.apply(register_hook)
+        # make a forward pass
+        self(x)
+        # remove these hooks
+        for h in hooks:
+            h.remove()
+
+        # print out neatly
+        names = list(self._modules.keys())
```

Comment:
This only prints the modules that are in the parent module, but what if you have something like the following?
```python
class Model0(nn.Module):
    def __init__(self):
        super(Model0, self).__init__()
        self.conv1 = nn.Conv2d(1, 1, 1)
    def forward(self, input):
        return self.conv1(input)

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.conv1 = nn.Conv2d(1, 1, 1)
        self.mod = Model0()
    def forward(self, input):
        return self.mod(self.conv1(input))
```
Given that you are applying the hooks to all modules iteratively (because of the `apply`), I think we should also walk recursively here, maybe using `self.named_modules()`?

Reply:
I believe [`apply`](http://pytorch.org/docs/master/nn.html#torch.nn.Module.apply) goes recursively to submodules? I think this is not an issue.